pandas/core/generic.py

```diff
@@ -2012,6 +2013,15 @@ def sample(self, n=None, frac=None, replace=False, weights=None, random_state=No
         # Check weights for compliance
         if weights is not None:
 
+            # If a series, align with frame
+            if isinstance(weights, pd.Series):
+                obj_index = self.axes[axis]
+
+                if len(obj_index.difference(weights.index)) != 0:
```

Comment:
why are you then reindexing if you don't allow these do differ? I would just reindex and be done. Most pandas functions work like this.